lgc/patch/MeshTaskShader.cpp

```diff
@@ -820,6 +844,152 @@ void MeshTaskShader::writeTaskPayload(Value *writeValue, Value *byteOffset) {
+// Create a task payload atomic operation other than compare-and-swap. Result type is the same as the input value type.
+//
+// @param atomicOp : Atomic op to perform
+// @param ordering : Atomic ordering
+// @param inputValue : Input value
+// @param byteOffset : Byte offset within the payload structure
+// @returns : Original value read from the task payload
+Value *MeshTaskShader::taskPayloadAtomic(unsigned atomicOp, AtomicOrdering ordering, Value *inputValue,
+                                         Value *byteOffset) {
+  auto entryPoint = m_builder->GetInsertBlock()->getParent();
+  assert(getShaderStage(entryPoint) == ShaderStageTask);
+
+  assert(inputValue->getType()->isIntegerTy() || inputValue->getType()->isFloatingPointTy());
+
+  auto payloadRingBufDesc = m_pipelineSysValues.get(entryPoint)->getTaskPayloadRingBufDesc();
+  auto payloadRingEntryOffset = getPayloadRingEntryOffset(entryPoint);
+
+  // NOTE: buffer.atomic.swap.f64 is not supported in LLVM backend, so we convert double to int64.
+  bool doubleToInt64 = atomicOp == AtomicRMWInst::Xchg && inputValue->getType()->isDoubleTy();
+  if (doubleToInt64)
+    inputValue = m_builder->CreateBitCast(inputValue, m_builder->getInt64Ty());
+
+  Intrinsic::ID intrinsic = Intrinsic::not_intrinsic;
+  switch (atomicOp) {
+  case AtomicRMWInst::Xchg:
+    intrinsic = Intrinsic::amdgcn_raw_buffer_atomic_swap;
+    break;
+  case AtomicRMWInst::Add:
+    intrinsic = Intrinsic::amdgcn_raw_buffer_atomic_add;
+    break;
+  case AtomicRMWInst::Sub:
+    intrinsic = Intrinsic::amdgcn_raw_buffer_atomic_sub;
+    break;
+  case AtomicRMWInst::And:
+    intrinsic = Intrinsic::amdgcn_raw_buffer_atomic_and;
+    break;
+  case AtomicRMWInst::Or:
+    intrinsic = Intrinsic::amdgcn_raw_buffer_atomic_or;
+    break;
+  case AtomicRMWInst::Xor:
+    intrinsic = Intrinsic::amdgcn_raw_buffer_atomic_xor;
+    break;
+  case AtomicRMWInst::Max:
+    intrinsic = Intrinsic::amdgcn_raw_buffer_atomic_smax;
+    break;
+  case AtomicRMWInst::Min:
+    intrinsic = Intrinsic::amdgcn_raw_buffer_atomic_smin;
+    break;
+  case AtomicRMWInst::UMax:
+    intrinsic = Intrinsic::amdgcn_raw_buffer_atomic_umax;
+    break;
+  case AtomicRMWInst::UMin:
+    intrinsic = Intrinsic::amdgcn_raw_buffer_atomic_umin;
+    break;
+  case AtomicRMWInst::FAdd:
+    intrinsic = Intrinsic::amdgcn_raw_buffer_atomic_fadd;
+    break;
+  case AtomicRMWInst::FMax:
+    intrinsic = Intrinsic::amdgcn_raw_buffer_atomic_fmax;
+    break;
+  case AtomicRMWInst::FMin:
+    intrinsic = Intrinsic::amdgcn_raw_buffer_atomic_fmin;
+    break;
+  default:
+    llvm_unreachable("Unexpected atomic operation!");
+    break;
+  }
+
+  if (ordering == AtomicOrdering::Release || ordering == AtomicOrdering::AcquireRelease ||
+      ordering == AtomicOrdering::SequentiallyConsistent)
+    m_builder->CreateFence(AtomicOrdering::Release, SyncScope::System);
```

Comment:
If this instruction is used in the task shader, then Workgroup scope should be sufficient: during the task shader, only threads from the same task shader workgroup can access the payload. This applies to all the fences for both instruction types (atomicrmw and cmpswap).

In the mesh shader, we do need system scope (well, technically device scope) because mesh shader workgroups for the same task payload can be distributed everywhere on the device.

Reply:
Thank you. I will correct this. We only need to handle task payload in task shader. For mesh shader, task payload is read-only. And all atomic operations are not applicable.